signac/contrib/_index.py

```diff
@@ -0,0 +1,299 @@
+        if _id is not None:
+            reduce_results({_id} if _id in self else set())
+
+        # Extract all logical-operator expressions for now.
+        or_expressions = expr.pop("$or", None)
+        and_expressions = expr.pop("$and", None)
+        not_expression = expr.pop("$not", None)
+
+        # Reduce the result based on the remaining non-logical expression:
+        for key, value in _nested_dicts_to_dotted_keys(expr):
+            reduce_results(self._find_expression(key, value))
+            if not result_ids:
+                # No matches, so exit early.
+                return set()
+
+        # Reduce the result based on the logical-operator expressions:
+        if not_expression is not None:
+            not_match = self._find_result(not_expression)
+            reduce_results(set(self).difference(not_match))
+            if not result_ids:
+                # No matches, so exit early.
+                return set()
+
+        if and_expressions is not None:
+            _check_logical_operator_argument("$and", and_expressions)
+            for expr_ in and_expressions:
+                reduce_results(self._find_result(expr_))
+                if not result_ids:
+                    # No matches, so exit early.
+                    return set()
+
+        if or_expressions is not None:
+            _check_logical_operator_argument("$or", or_expressions)
+            or_results = set()
+            for expr_ in or_expressions:
+                or_results.update(self._find_result(expr_))
+            reduce_results(or_results)
+
+        assert result_ids is not None
+        return result_ids
+
+    def find(self, filter=None):
+        """Return a result vector of ids for the given filter and limit.
+
+        This function normalizes the filter argument and then attempts to
+        build a result vector for the given key-value queries.
+        For each key that is queried, an internal index is built and then
+        searched.
+
+        The results are a set of ids, where each id is the value of the
+        primary key of a document that matches the given filter.
+
+        The find() method uses the following optimizations:
+
+            1. If the filter is None, the result is directly returned, since
+               all documents will match an empty filter.
+            2. The filter is processed key by key, once the result vector is
+               empty it is immediately returned.
+
+        Parameters
+        ----------
+        filter : dict
+            The filter argument that all documents must match (Default value = None).
+
+        Returns
+        -------
+        set
+            A set of ids of documents that match the given filter.
+
+        Raises
+        ------
+        ValueError
+            When the filter argument is invalid.
+
+        """
+        if not filter:
+            return set(self)
+
+        filter = json.loads(json.dumps(filter))  # Normalize
```

Comment:
This encode/decode blocks a large class of invalid inputs. All valid queries are JSON-encodable dicts, and this ensures we don't have arbitrary Python objects being passed in to queries. We have a second validation step in `_valid_filter` that ensures we block cases like `[]`. However, I think that helper function could be eliminated: it appears that the only invalid filter is a top-level list, `[]`. Doing a round-trip through JSON encoding and ensuring that the filter is a `dict` should be enough for validation.

Reply:
Right, but at this stage haven't we always already parsed the filter using `filterparse.parse_filter` to validate the inputs? I'm pretty sure that since we are no longer exposing this API publicly we can make more assumptions about the validity of the inputs than we are currently making. And JSON encoding is quite expensive.

Reply:
Here's my quick analysis. We call `find` in schema.py::_build_job_statepoint_index with no arguments. That gets used in `import_export.py` and in `Project.detect_schema`. We also call `_Index.find` in `Project._find_job_ids`, where the filter has already been parsed by `parse_filter`. I think that's all the calls to this function that exist. So I think we don't need this extra normalization at all.

Reply:
Yes, the purpose is input validation and normalization and it certainly fixed related bugs when first introduced. I would recommend that in addition to @vyasr ' analysis, we ensure that this is covered by unit tests for unsanitized input before removing or replacing it.